sources/models/scalar-derive/src/lib.rs

```diff
@@ -0,0 +1,493 @@
+                    fn try_from(input: &#inner_ref_type) -> Result<Self, ValidationError> {
+                        Self::new(input)
+                    }
+                }
+
+                impl PartialEq<#inner_ref_type> for #scalar {
+                    fn eq(&self, other: &#inner_ref_type) -> bool {
+                        let self_as_str: &str = self.as_ref();
+                        let other_as_str: &str = other.as_ref();
+                        self_as_str.eq(other_as_str)
+                    }
+                }
+
+                impl PartialEq<#scalar> for #inner_ref_type {
+                    fn eq(&self, other: &#scalar) -> bool {
+                        let self_as_str: &str = self.as_ref();
+                        let other_as_str: &str = other.as_ref();
+                        self_as_str.eq(other_as_str)
+                    }
+                }
+
+                impl PartialEq<#scalar> for &#inner_ref_type {
+                    fn eq(&self, other: &#scalar) -> bool {
+                        let other_as_str: &str = other.inner().as_ref();
+                        let self_as_str: &str = self.as_ref();
+                        self_as_str.eq(other_as_str)
+                    }
+                }
+
+                impl PartialEq<&#inner_ref_type> for #scalar {
+                    fn eq(&self, other: &&#inner_ref_type) -> bool {
+                        let other_as_str: &str = (*other);
+                        let self_as_str: &str = self.as_ref();
+                        self_as_str.eq(other_as_str)
+                    }
+                }
+            );
+            stream.append_all(code.into_iter());
+        }
+
+        // If the inner type is String, then we already have this implemented. If not, we can add
+        // this as a convenience since we know the inner type implements `Display`.
+        if !is_string(&self.inner_type) {
+            let code = quote!(
+                impl From<#scalar> for String {
+                    fn from(scalar: #scalar) -> Self {
+                        scalar.unwrap().to_string()
+                    }
+                }
+            );
+            stream.append_all(code.into_iter());
+        }
+    }
+}
+
+/// Given a type `t`, what is its `typename`?
+fn typename(t: &syn::Type) -> String {
+    let mut stream = proc_macro2::TokenStream::new();
+    t.to_tokens(&mut stream);
+    stream.to_string()
+}
+
+/// Whether or not the type assumed to be `std::string::String`
+fn is_string(t: &str) -> bool {
+    t == "String" || t == "string::String" || t == "std::string::String"
+}
+
+/// This function finds the "inner" field of the struct. The most common example is:
+///
+/// ```no_run
+/// struct SomeStruct {
+///     inner: String
+/// }
+/// ```
+///
+/// In the above example, this function would return the field ("inner", "String")
+fn find_inner_field(data_struct: DataStruct, field_name: Option<&str>) -> (String, String) {
```

Comment:
It might be nice to write some test cases to validate "correct usage." I'm not sure that there's a good way to validate errors in proc macros?

Reply:
> It might be nice to write some test cases to validate "correct usage." I'm not sure that there's a good way to validate errors in proc macros?

The doc tests are compiling and running. Do you think that's good enough or should I add some more tests? These can be added in doc tests, a `tests` directory, or an `examples` directory. They can't be in the same crate like unit tests are.

Testing failure cases would be pretty difficult. I'm not sure you can mark a test as "shouldn't compile". The good news is that a panic in this code can only ever break a build, it can't affect Bottlerocket runtime.